### ISSUE TITLE
Need an option to not save a local user in Keycloak when using an IdP

### ISSUE THREAD

**Issue by annasarte** [keycloak/keycloak#13739]
### Description

Currently, if you are using an Identity Provider (in my case an OIDC compliant provider such as Okta) in Keycloak for authentication, the user is always imported into Keycloak and is persisted in the keycloak database so it will appear as a local user.  With GDPR, we have customers who do NOT want any of their users stored anywhere except in their own systems. At most, we can possibly save a "user" with minimum info but the info has to be anonymous data (like an "ID" that is not PII -- so no names, emails, etc). I came across this old post [https://issues.redhat.com/browse/KEYCLOAK-4429](https://issues.redhat.com/browse/KEYCLOAK-4429) and that is complaining about the same issue I am trying to solve. The last post on that says it was deferred.  Any new updates on that?   I am on Keycloak 18.0.2 and this issue is still happening. Please advise and thanks in advance.

### Discussion

https://issues.redhat.com/browse/KEYCLOAK-4429

### Motivation

GDPR - customer owns this data and not us.

**Comment by cesdperez** [comment]
We need exactly the same thing, also related to GDPR.

I also found #11334 with no answers.

Is it at least possible to achieve this with a custom authenticator?

**Comment by sschu** [comment]
AFAIK it could be possible with a custom user store that only stores user data in memory. However, I am not sure this makes a difference according to GDPR. If you want Keycloak to generate a token containing personal data, it will process and at least temporarily store that data...

**Comment by annasarte** [comment]
Hi.  I tried a few things with no luck:

- I tried creating my own 'first broker login' flow and it basically just had the first step of reviewing the profile and that was it.  The original 'first broker login' flow had steps following that related to account creation.  But if I used my new flow in my IdP configuration under the field "first login flow" I can see that the user gets authenticated up at the IdP level (I see authentication success happen in Okta) but in keycloak I cannot login and get an error that says can't authenticate.

- I also tried starting keycloak using ----storage-area-user=chm (per link  [https://www.keycloak.org/server/all-config#_storage_experimental](https://www.keycloak.org/server/all-config#_storage_experimental) to see if it would make a difference but it was still saving it to the database.

Does anyone know:

- why Keycloak needs to persist user data? We will be ok with it persisting minimum data like a non personally identifiable ID) but not email, username, name etc.
- has anyone been successful with "chm" as a user store and what can I expect from this configuration?
- And @sschu have you been able to create your own custom store?

Any help is appreciated.

**Comment by sschu** [comment]
@annasarte No, we have not implemented this solution. It might get on our roadmap but currently, we don't need it.

**Comment by annasarte** [comment]
To Keycloak folks -- any update on this since https://issues.redhat.com/browse/KEYCLOAK-4429 is closed without a resolution ?

**Comment by rep-movsd** [comment]
Any update on this issue?

**Comment by ammad27k** [comment]
any update, we need this

**Comment by fredericoschardong** [comment]
Has anyone found a solution for this issue?

**Comment by sschu** [comment]
@fredericoschardong There is already an experimental feature for that (https://github.com/keycloak/keycloak/pull/23977) which is currently being extended (see e.g. https://github.com/keycloak/keycloak/pull/23977).

**Comment by jaro-s** [comment]
Is there a timeline for moving the TRANSIENT_USERS feature from experimental to production-ready?

**Comment by sharpedavid** [comment]
If you prefer not to store first name, last name, and email, you can use Keycloak mappers (or create custom ones) to replace these values. Alternatively, you can configure the external IDP to exclude them altogether. That said, many systems store these common attributes for legitimate purposes while complying with GDPR—so it may also be worth considering how to handle them within regulatory guidelines.

I wouldn’t be surprised if Keycloak eventually supports truly 'transient' users without storing anything. However, that seems quite different from its current design, which relies on persistent records to support features like roles and attributes. Given that, I’m not holding my breath.
